File: tri.js
"use strict";
// ТРЕУГОЛЬНИК — монитор треугольного арбитража на битке (MEXC фьючерсы, read-only).
// USDT↔BTC↔USDC↔USDT через BTC_USDT · BTC_USDC · USDC_USDT.
// Показывает живой gross/net (тейкер) и потолок для мейкер-исполнения. Ордера НЕ шлёт.
(function(){
  const g=(id)=>document.getElementById(id);
  const T={ timer:null };

  const css=document.createElement("style"); css.textContent=`
  #triwin{display:flex;flex-direction:column;font:12px ui-monospace,Consolas,monospace}
  #tribody{flex:1;overflow-y:auto;padding:10px 12px}
  .trihead{display:flex;gap:8px;align-items:center;color:#8a929c;font-size:10px;margin-bottom:8px;flex-wrap:wrap}
  .trihead .chip{border-radius:4px;padding:1px 6px;background:rgba(255,255,255,.05)}
  .trihero{display:flex;align-items:baseline;gap:10px;margin:6px 0 12px}
  .trihero .big{font-size:34px;font-weight:800;letter-spacing:-1px}
  .trihero .big.pos{color:#4be38a}.trihero .big.neg{color:#ff8a86}
  .trihero .sub{font-size:11px;color:#8a929c}
  .trirow{display:grid;grid-template-columns:150px 1fr 1fr;gap:6px;padding:4px 0;border-bottom:1px solid rgba(255,255,255,.05)}
  .trirow .k{color:#8a929c}
  .trirow .pos{color:#4be38a}.trirow .neg{color:#ff8a86}
  .trilegs{margin-top:12px}
  .trileg{display:grid;grid-template-columns:110px 1fr 1fr 70px;gap:6px;padding:3px 0;font-size:11px}
  .trileg .sym{color:#dfe6ee;font-weight:700}
  .trileg .lbl{color:#5b6573}
  .trispark{display:flex;align-items:flex-end;gap:1px;height:38px;margin:8px 0}
  .trispark i{flex:1;background:#2e5;opacity:.5;min-height:1px}
  .trispark i.neg{background:#e55}
  .trinote{margin-top:12px;padding:8px;border:1px solid rgba(230,169,67,.25);border-radius:6px;background:rgba(230,169,67,.06);color:#d9c08a;font-size:10px;line-height:1.5}
  .triwin-off{color:#5b6573;padding:20px;text-align:center}`;
  document.head.appendChild(css);

  const win=document.createElement("div");
  win.id="triwin"; win.className="wbwin hidden";
  win.style.cssText="left:auto;right:24px;top:120px;width:460px;height:520px";
  win.innerHTML=`
    <div class="wbtitle" id="tridrag">
      <span class="wbcoin">🔺 ТРЕУГОЛЬНИК BTC</span>
      <span class="wbsp"></span>
      <span id="tri-state" style="font-size:9px;color:#5b6573"></span>
      <button class="wbx" id="triclose">×</button>
    </div>
    <div id="tribody"><div class="triwin-off">Подключаюсь к MEXC…</div></div>
    <div class="wbresize" id="trires"></div>`;
  document.body.appendChild(win);

  const bp=(x)=> (x>=0?"+":"")+x.toFixed(2)+" bps";
  const cls=(x)=> x>=0?"pos":"neg";

  function render(s){
    const body=g("tribody");
    if(!s||!s.ok){ body.innerHTML='<div class="triwin-off">Нет данных с MEXC…</div>'; return; }
    const best=s.best_net, dir=s.best_dir;
    const winrate = s.cycles? Math.round(100*s.windows/s.cycles):0;
    // мейкер-потолок: две BTC-ноги 0%, петля ~1bps → net_maker ≈ gross - 0.5..1 bps
    const netMaker = s.best_gross - 1.0;
    const B=s.books||{};
    const leg=(sym,label)=>{ const b=B[sym]; if(!b) return "";
      return `<div class="trileg"><span class="sym">${sym}</span>`+
             `<span class="lbl">bid ${b.bid}</span><span class="lbl">ask ${b.ask}</span>`+
             `<span class="lbl">${label}</span></div>`; };
    const hist=s.hist||[]; const mx=Math.max(1,...hist.map(Math.abs));
    const spark=hist.map(v=>`<i class="${v<0?'neg':''}" style="height:${Math.max(2,Math.abs(v)/mx*100)}%;opacity:${v>0?.85:.5}"></i>`).join("");
    body.innerHTML=`
      <div class="trihead">
        <span class="chip">циклов ${s.cycles}</span>
        <span class="chip">окон (net&gt;0): ${s.windows} · ${winrate}%</span>
        <span class="chip">USDC/USDT спред ${s.usdc_spread_bps.toFixed(2)} bps</span>
      </div>
      <div class="trihero">
        <span class="big ${cls(best)}">${bp(best)}</span>
        <span class="sub">лучший net (тейкер), направление ${dir}<br>gross ${bp(s.best_gross)} · мейкер-потолок ≈ ${bp(netMaker)}</span>
      </div>
      <div class="trispark">${spark}</div>
      <div class="trirow"><span class="k">Цикл A · USDT→BTC→USDC→USDT</span><span class="${cls(s.gross_a)}">gross ${bp(s.gross_a)}</span><span class="${cls(s.net_a)}">net ${bp(s.net_a)}</span></div>
      <div class="trirow"><span class="k">Цикл B · USDT→USDC→BTC→USDT</span><span class="${cls(s.gross_b)}">gross ${bp(s.gross_b)}</span><span class="${cls(s.net_b)}">net ${bp(s.net_b)}</span></div>
      <div class="trilegs">
        ${leg("BTC_USDT","maker 0%")}
        ${leg("BTC_USDC","maker 0%")}
        ${leg("USDC_USDT","maker 0.01%")}
      </div>
      <div class="trinote">Ядро края = лонг BTC_USDC + шорт BTC_USDT (обе ноги maker 0% → бесплатно).
      «gross» = потолок при мейкер-исполнении. «net» = если шмальнуть 3 маркета (тейкер, обычно в минус).
      Барьер: получить мейкер-филл на обеих BTC-ногах до схлопывания разрыва. Монитор, ордера не шлёт.</div>`;
  }

  async function poll(){
    try{ const r=await fetch("/api/tri/state"); const s=await r.json(); render(s);
      const st=g("tri-state"); if(st) st.textContent = s&&s.ok? ("обновлено "+new Date(s.ts*1000).toLocaleTimeString()) : (s&&s.err? "ошибка" : "…");
    }catch(e){}
  }
  function open(){ win.classList.remove("hidden"); win.style.zIndex=46; poll();
    if(!T.timer) T.timer=setInterval(poll, 3000); }
  function close(){ win.classList.add("hidden"); if(T.timer){ clearInterval(T.timer); T.timer=null; } }

  const btn=g("tribtn"); if(btn) btn.onclick=()=>{ win.classList.contains("hidden")?open():close(); };
  g("triclose").onclick=close;
  // перетаскивание/ресайз — как у прочих окон (dock.js вешает по классам wbtitle/wbresize)
})();

File: tri.py
"""ТРЕУГОЛЬНИК: монитор треугольного арбитража на битке — фьючерсы MEXC (read-only).

Узлы: USDT — USDC — BTC (биток в центре). Три реальных перпа на одной бирже:
    BTC_USDT · BTC_USDC · USDC_USDT
Цикл A: USDT→BTC→USDC→USDT.   Цикл B: USDT→USDC→BTC→USDT (зеркало).
Край = когда BTC_USDC/BTC_USDT ≠ прямому курсу USDC_USDT.

Считаем по РЕАЛЬНЫМ bid/ask (на каждой ноге пересекаем спред) минус комиссии.
На MEXC maker=0% → в мониторе показываем и тейкер-сценарий (шмальнуть 3 маркета),
и «gross» (без комиссий = потолок для мейкер-исполнения лимитками).

ТОЛЬКО МОНИТОР: ордера НЕ шлёт. Даёт состояние во фронт через server.py /api/tri/state.
"""
import json
import threading
import time
import urllib.request
from collections import deque

FUT = "https://contract.mexc.com/api/v1"
BTC_USDT, BTC_USDC, USDC_USDT = "BTC_USDT", "BTC_USDC", "USDC_USDT"
SYMS = (BTC_USDT, BTC_USDC, USDC_USDT)

CFG = {"poll_sec": 4.0}
FEES = {BTC_USDT: 0.0002, BTC_USDC: 0.0002, USDC_USDT: 0.0004}  # taker, перепроверяем из API

_LOCK = threading.Lock()
_STATE = {
    "ok": False, "ts": 0.0, "err": "",
    "books": {}, "fees": dict(FEES),
    "gross_a": 0.0, "net_a": 0.0, "gross_b": 0.0, "net_b": 0.0,
    "best_net": 0.0, "best_dir": "-", "best_gross": 0.0,
    "usdc_spread_bps": 0.0, "windows": 0, "cycles": 0,
    "hist": [],  # последние N best_net (bps) для мини-графика
}
_HIST = deque(maxlen=120)


def _get(url, timeout=15):
    req = urllib.request.Request(url, headers={"User-Agent": "squad-tri"})
    return json.loads(urllib.request.urlopen(req, timeout=timeout).read().decode())


def _verify_fees():
    try:
        data = _get(f"{FUT}/contract/detail")
        arr = data.get("data") if isinstance(data, dict) else data
        for it in arr or []:
            if it.get("symbol") in FEES:
                FEES[it["symbol"]] = float(it.get("takerFeeRate", FEES[it["symbol"]]))
    except Exception:  # noqa: BLE001 — не критично, остаёмся на дефолтах
        pass


def _fetch_books():
    data = _get(f"{FUT}/contract/ticker")
    arr = data.get("data") if isinstance(data, dict) else data
    out = {}
    for d in arr or []:
        s = d.get("symbol")
        if s in SYMS and d.get("bid1") and d.get("ask1"):
            out[s] = (float(d["bid1"]), float(d["ask1"]))
    return out


def _cycle_a(books, use_fee):
    a_btcusdt = books[BTC_USDT][1]     # купить BTC по ask
    b_btcusdc = books[BTC_USDC][0]     # продать BTC по bid
    b_usdcusdt = books[USDC_USDT][0]   # продать USDC по bid
    f = FEES if use_fee else {s: 0.0 for s in SYMS}
    btc = (1.0 / a_btcusdt) * (1 - f[BTC_USDT])
    usdc = btc * b_btcusdc * (1 - f[BTC_USDC])
    usdt = usdc * b_usdcusdt * (1 - f[USDC_USDT])
    return (usdt - 1.0) * 10000


def _cycle_b(books, use_fee):
    a_usdcusdt = books[USDC_USDT][1]   # купить USDC по ask
    a_btcusdc = books[BTC_USDC][1]     # купить BTC по ask
    b_btcusdt = books[BTC_USDT][0]     # продать BTC по bid
    f = FEES if use_fee else {s: 0.0 for s in SYMS}
    usdc = (1.0 / a_usdcusdt) * (1 - f[USDC_USDT])
    btc = (usdc / a_btcusdc) * (1 - f[BTC_USDC])
    usdt = btc * b_btcusdt * (1 - f[BTC_USDT])
    return (usdt - 1.0) * 10000


def state():
    with _LOCK:
        s = dict(_STATE)
        s["hist"] = list(_HIST)
        return s


def set_cfg(body):
    try:
        v = float(body.get("poll_sec", CFG["poll_sec"]))
        CFG["poll_sec"] = max(2.0, min(30.0, v))
    except Exception:  # noqa: BLE001
        pass
    return {"ok": True, "cfg": dict(CFG)}


def poller():
    """Фоновый цикл: тянет 3 стакана, считает край, копит статистику."""
    _verify_fees()
    windows = 0
    cycles = 0
    while True:
        try:
            books = _fetch_books()
            if len(books) < 3:
                time.sleep(CFG["poll_sec"])
                continue
            cycles += 1
            ga, na = _cycle_a(books, False), _cycle_a(books, True)
            gb, nb = _cycle_b(books, False), _cycle_b(books, True)
            best_net = max(na, nb)
            best_dir = "B" if nb >= na else "A"
            best_gross = gb if best_dir == "B" else ga
            if best_net > 0:
                windows += 1
            ub, ua = books[USDC_USDT]
            usdc_spread = (ua - ub) / ub * 10000
            _HIST.append(round(best_net, 2))
            with _LOCK:
                _STATE.update({
                    "ok": True, "ts": time.time(), "err": "",
                    "books": {k: {"bid": v[0], "ask": v[1]} for k, v in books.items()},
                    "fees": dict(FEES),
                    "gross_a": ga, "net_a": na, "gross_b": gb, "net_b": nb,
                    "best_net": best_net, "best_dir": best_dir, "best_gross": best_gross,
                    "usdc_spread_bps": usdc_spread, "windows": windows, "cycles": cycles,
                })
        except Exception as exc:  # noqa: BLE001 — монитор не должен падать
            with _LOCK:
                _STATE["err"] = repr(exc)[:200]
        time.sleep(CFG["poll_sec"])
